### ISSUE TITLE
question: relationship between quantity of parameters and buffers of a model, transforms values of a datapoint during training and reconstruction results or hyperparameters values choice

### ISSUE THREAD

**Issue by cortexmancer** [JonasGeiping/breaching#34]
Hello. I hope you're doing fine.

I got really fixated into archiving results while running my experiments and I got a little out of the road with the main objective that was constructing the exact reference to the checkpoint to collect data from clients. And then I have run into some results, with "Beyond Inferring" attack and made some comparisons and hypothesis that I do not have the means to prove, and I would like to share with you to see your opinion.

I started to use the "overview" function that the simulation on Breaching uses, and that provided me some insight into things. For the following results and some outputs that I'll be sharing, I've used the following models: "Net" the default convnet on flower quick-start, "ConvNet" with BN, "ConvNetSmall" without BN and ConvNet Beyond.

So to a start, the model architecture that "overview" prints with simulation for each model:

Net:
``` shell
Model architecture loaded with 62,006 parameters and 0 buffers.
Overall this is a data ratio of      20:1 
for target shape [1, 3, 32, 32] given that num_queries=1.
```

ConvNet with width 32:
``` shell
Model architecture loaded with 733,642 parameters and 1,608 buffers.
Overall this is a data ratio of     239:1 
for target shape [1, 3, 32, 32] given that num_queries=1.
```

ConvNet with width 64:
``` shell
Model architecture loaded with 2,904,970 parameters and 3,208 buffers.
Overall this is a data ratio of     946:1 
for target shape [1, 3, 32, 32] given that num_queries=1.
```

ConvNetSmall with width 256:
``` shell
Model architecture loaded with 15,355,402 parameters and 0 buffers.
Overall this is a data ratio of    4999:1 
for target shape [1, 3, 32, 32] given that num_queries=1.
```

ConvNetBeyond:
``` shell
Model architecture loaded with 206,047,306 parameters and 0 buffers.
Overall this is a data ratio of   67073:1 
for target shape [1, 3, 32, 32] given that num_queries=1.
```

Well, for those architectures there is clearly a progression with the size of the parameters that a model load/has, and only "ConvNet" is capable of loading buffers data.

Question 1: Only models with BN layers are able to load buffers datas?
This maybe can be answers with a little peak at PyTorch documentation and I'll be looking into it. 

``` python
    target_information = 1 * torch.as_tensor(data_cfg_default.shape).prod()
    logging.info(
        f"\nModel architecture loaded with {num_params:,} parameters and {num_buffers:,} buffers.\n"
        f"Overall this is a data ratio of { 1 * num_params / target_information:7.0f}:1 \n"
        f"for target shape {[1, *data_cfg_default.shape]} given that num_queries=1.\n"
    )
```

I have hardcoded num_datapoints to 1, but I would like to understand what "data ratio" means.
It would be the amount of data that an given datapoint can store in the parameters layers?

Now checkout some of the results:
For each image, to left to right:
1: Default Total variation scale (0.2352)
2: 1e-4 (0.0001)
3: 1e-4 (0.001)
4: Original image

Net:
<img width="963" height="231" alt="Image" src="https://github.com/user-attachments/assets/a18552cc-6694-4d5d-a21c-470b4e6f87cd" />

ConvNet with width 32:
<img width="956" height="231" alt="Image" src="https://github.com/user-attachments/assets/7748d041-6787-4453-b809-508f497c4393" />

ConvNet with width 64:
<img width="975" height="247" alt="Image" src="https://github.com/user-attachments/assets/7862c409-abbc-49b4-825c-e768d28e0780" />

ConvNetSmall:
<img width="958" height="234" alt="Image" src="https://github.com/user-attachments/assets/20d3c125-e439-4970-bcee-f6fc6707b991" />

ConvNetBeyond:
<img width="963" height="237" alt="Image" src="https://github.com/user-attachments/assets/1c1e404b-d933-4fe2-b5ca-4f2a2b2ea7c9" />

- So clearly when the number of parameters increased, the results of the reconstruction became better overall.
- Net worked better with default hyperameter for scale and got a lot of noise when other values was used
- ConvNet the model used in breaching examples for Beyond Inferring got really poor results, even with scale 1e-4, and this is really strange since simulation seen to reconstruct pretty well when simulating. I don't know if this is because I am collecting the parameters or the public buffers in a wrong position, but for this I think I am doing right this time. 
Question 2: Would you have any ideas for the reason for this behavior?

Then, I noticed that the transforms that flower uses by default on the training are different from the ones that Breaching uses (0.5 for all channels in mean and std). I was trying to approximate the simulation with my gradient collection setup, so I did another run, using the values that Breaching uses for mean and std.

Then I got the following results:

Net:
<img width="950" height="223" alt="Image" src="https://github.com/user-attachments/assets/231e529e-70ab-4ab5-b26d-8a271cf0e308" />

ConvNet W32:
<img width="973" height="239" alt="Image" src="https://github.com/user-attachments/assets/1ea637bf-be04-49a0-9bbe-dc6e902f9b69" />

ConvNet W64:
<img width="963" height="218" alt="Image" src="https://github.com/user-attachments/assets/ad6ea557-e850-44af-a5f1-538649e61058" />

ConvNet Small:
<img width="965" height="231" alt="Image" src="https://github.com/user-attachments/assets/54027ac2-9667-4d47-9831-087ef056de00" />

ConvNet Beyond:
<img width="959" height="245" alt="Image" src="https://github.com/user-attachments/assets/9a02901a-59ba-45f2-9110-a9bbab7e22a1" />

While ConvNet 64 got an result closer to the simulation, all other results became worse, and ConvNet Beyond performed more or less in the same way with both transformations. I have concluded that the transformations that Breaching uses by default makes the reconstruction harder overall. 

Question 3: It is feasible to say that the transformation that an data has on train affects the result?
Question 4: Can this transformation affect the choice of a value for an hyperpameter?
Question 5: Can the choice for a hyperparameter value be rationalized? It is clear to me that you choose it based in dataset shape because some commentaries in the examples, so maybe I am lacking the understatement of what this hyperparameter do, but I think is about the distribution of the pixels. 

Finally in Beyond Inferring example you state: 

> Neither CIFAR-10 or this model are used in the original paper. The model from the original paper can be retrieved with case.model=convnet_beyond and has an (even) larger fully-connected part than the default ConvNet used here.

Question 6: So would an reconstruction result, for Beyond Inferring specific, or for optimization attacks in general be associated with the size of the fully-connect part?
Question 7: Why did you used an model with BN as the default case for beyond inferring, that can reconstruct only batchs of size 1? Could this just because this model is used in some other more complex cases? Or to create an more difficult scenario? 

With all this, I felt the lack that I have in understating the models implementation... And the need of explain this in my work, this is my next step of learning. I feel like I have jumped some steps of the ladder, and now I have to go back a little to provide a reasonable explanation for those implementations. 

It is clearly to me as well that an ideal setup for training may be far from the ideal setup for an reconstruction. So an wise choose in the model implementation in a real scenario could avoid the breach of privacy in real scenario, in a way of "less is more" maybe? 

Thank you for your time and attention!
